### DISCUSSION TITLE
Real time trigger processing

### DISCUSSION THREAD

**Discussion by gilberto-p-matos-alb** [jobrunr/jobrunr#442] [Ideas]
Hi, first of all, congrats on developing Jobrunr.

We've been using quartz for several years as our base for distributed scheduling and one of the main pains we always struggled with was its triggering per second rate, which was much lower than expected because the process of getting the triggers and firing them is very time demanding
* Also, our intended persistence storage was MongoDB, which isn't supported, leading to other issues

Because of that we recently started testing other alternatives and JobRunr has provided great results, providing a high scheduling trigger rate.
Also, the visibility given by the provided dashboard and its native support for MongoDB / NoSql is a plus.

One "caveat" we found, is that the query executed to fetch the scheduled jobs takes into consideration the polling interval time, meaning that scheduled tasks may be fired a few seconds ahead of time (the same happens in quartz and other frameworks). We understand the rationale behind this approach, and for long-run tasks, this isn't an issue.

Unfortunately, this is undesirable for our scenario because not only do we trigger a lot of tasks simultaneously, but we also have some aggressive timeouts (5 to 30 seconds), which we don't want to fire ahead of time.
* If we configure the poll interval with a minimum of 5 seconds, some of these triggers may fire instantly

We took a look into Jobrunr code, mainly the `JobZooKeeper` and `BackgroundJobServer` classes.
While reading through `BackgroundJobServer`, we noticed that the only implementation of `JobRunrExecutor` is `ScheduledThreadPoolJobRunrExecutor`. Because it is an extension of java's `ScheduledThreadPoolExecutor`, it can schedule tasks.
So we did a little _hack_, where we tried the following change to `BackgroundJobServer`

```java
    void processJob(Job job) {
        BackgroundJobPerformer backgroundJobPerformer = new BackgroundJobPerformer(this, job);

        // fetch the last scheduled state, if any
        ScheduledState scheduledState = job.getJobStates()
                .stream()
                .filter(j -> j instanceof ScheduledState)
                .map(j -> (ScheduledState) j)
                .max(Comparator.comparing(ScheduledState::getScheduledAt))
                .orElse(null);
        long now = System.currentTimeMillis();
        
        // if it's a scheduled task and its fire time hasn't been reached yet, then schedule it instead of execute it
        if (scheduledState != null && scheduledState.getScheduledAt().toEpochMilli() > now) {
            long timeUntilProcess = scheduledState.getScheduledAt().toEpochMilli() - now;
            ((ScheduledThreadPoolJobRunrExecutor) jobExecutor).schedule(backgroundJobPerformer, timeUntilProcess, TimeUnit.MILLISECONDS);
        } else {
            jobExecutor.execute(backgroundJobPerformer);
        }
        LOGGER.debug("Submitted BackgroundJobPerformer for job {} to executor service", job.getId());
    }
```

This is a very raw implementation, but our initial tests gave good results, with no triggers being fired ahead of time: with 2 nodes, we achieved a throughput of 350 tasks being triggered per second.
* We were able to achieve higher throughputs at the cost of having some delays

We were wondering if a similar approach was initially intended since the Executor implementation already is a ScheduledThreadPoolExecutor.
If not, would consider having this feature (for example, configuration of "fireAheadTimeMs", which by default was equal to pollIntervalInSec), or, as an alternative, provide a way of customizing this behaviour?

**Comment by rdehuyss** [comment]
Hi @gilberto-p-matos , although this was originally planned for JobRunrPro (exact scheduling) I'm also open for a PR for exact scheduling. 

With the implementation you suggest, I see some problems when having multiple `BackgroundJobServers` - it will lead to a lot of `ConcurrentJobModificationExceptions` impacting Job throughput.

If I would add exact job scheduling in the Pro version, would it warrant the license for you?

**Reply by gilberto-p-matos-alb** [reply]
Hi @rdehuyss, thank you for your reply.

Yes, we noticed there were more `ConcurrentJobModificationExceptions` being raised, leading to lower throughput. Also, other problems arise if the worker count is too low. This was only meant as a minimalistic proof of concept.
* It's worth mentioning, that even with more exceptions being raised, no data inconsistency was detected.

Regarding your question
> If I would add exact job scheduling in the Pro version, would it warrant the license for you?

I really can't answer you, we are merely developers :smile:, but I guess it would be easier to "sell" to our leaders that the cost of migration was that we needed to contribute to the framework's functionality, instead of paying for a licence.

That being said, if you are okay with it, we would like to contribute to Jobrunr and make this functionality available to all.
We have an idea of how to implement this, and if you're interested, we would like to discuss it with you.

**Reply by rdehuyss** [reply]
Could you add me to your LinkedIn network? It makes chatting a little bit easier :-). Thx!

**Reply by gilberto-p-matos-alb** [reply]
Invite sent :)
